src/realm.h

```diff
@@ -3576,6 +3554,78 @@ typedef struct realm_sync_error_compensating_write_info {
     realm_value_t primary_key;
 } realm_sync_error_compensating_write_info_t;
 
+/**
+ * Generic completion callback for asynchronous Realm User operations.
+ * @param userdata This must be the faithfully forwarded data parameter that was provided along with this callback.
+ * @param error Pointer to an error object if the operation failed, otherwise null if it completed successfully.
+ */
+typedef void (*realm_user_void_completion_func_t)(realm_userdata_t userdata, const realm_app_error_t* error);
+
+
+typedef const char* (*realm_user_get_access_token_cb_t)(realm_userdata_t userdata);
+typedef const char* (*realm_user_get_refresh_token_cb_t)(realm_userdata_t userdata);
+typedef realm_user_state_e (*realm_user_state_cb_t)(realm_userdata_t userdata);
+typedef bool (*realm_user_access_token_refresh_required_cb_t)(realm_userdata_t userdata);
+typedef realm_sync_manager_t* (*realm_user_get_sync_manager_cb_t)(realm_userdata_t userdata);
+typedef void (*realm_user_request_log_out_cb_t)(realm_userdata_t userdata);
+typedef void (*realm_user_request_refresh_user_cb_t)(realm_userdata_t userdata, realm_user_void_completion_func_t cb,
+                                                     realm_userdata_t cb_data);
+typedef void (*realm_user_request_refresh_location_cb_t)(realm_userdata_t userdata,
+                                                         realm_user_void_completion_func_t cb,
+                                                         realm_userdata_t cb_data);
+typedef void (*realm_user_request_access_token_cb_t)(realm_userdata_t userdata, realm_user_void_completion_func_t cb,
+                                                     realm_userdata_t cb_data);
+typedef void (*realm_user_track_realm_cb_t)(realm_userdata_t userdata, const char* path);
+typedef const char* (*realm_user_create_file_action_cb_t)(realm_userdata_t userdata, realm_sync_file_action_e action,
+                                                          const char* original_path,
+                                                          const char* requested_recovery_dir);
+
+
+/*
+ * Construct a SyncUser instance that uses SDK provided
+ * callbacks instead of core's User implementation. This type
+ * of user should not be used with core's App implementation.
+ */
+RLM_API realm_user_t* realm_user_new(
```

Comment:
This is very big, I see why you are doing this though, it is just one hop into core from managed code. And in 1 go you are setting all the callbacks. 
However, this is very difficult to maintain. Could we introduce a new realm struct, in which we have all the callbacks, and pass this struct around? In this way the SDK will have to specify whichever callback it needs, and we can easily add or remove callbacks .